File: Homework4/Hw4/Calculator.fs
module Hw4.Calculator

open System

type CalculatorOperation =
     | Plus = 0
     | Minus = 1
     | Multiply = 2
     | Divide = 3
     | Undefined = 4
     
let calculate (value1 : float) (operation : CalculatorOperation) (value2 : float) =
    match operation with
    | CalculatorOperation.Plus -> value1 + value2
    | CalculatorOperation.Minus -> value1 - value2
    | CalculatorOperation.Multiply -> value1 * value2
    | CalculatorOperation.Divide -> value1 / value2
    | _ -> ArgumentOutOfRangeException "Undefined operation" |> raise
    

File: Homework8/Hw8/Calculator/Calculator.cs
using Hw8.Common;

namespace Hw8.Calculator;

public class Calculator : ICalculator
{
    public double Calculate(double val1, Operation operation, double val2) =>
        operation switch
        {
            Operation.Plus => Plus(val1, val2),
            Operation.Minus => Minus(val1, val2),
            Operation.Multiply => Multiply(val1, val2),
            Operation.Divide => Divide(val1, val2),
            _ => throw new InvalidOperationException(Messages.InvalidOperationMessage)
        };

    public double Plus(double val1, double val2) => val1 + val2;
    public double Minus(double val1, double val2) => val1 - val2;

    public double Multiply(double val1, double val2) => val1 * val2;

    public double Divide(double val1, double val2) =>
        val2 < double.Epsilon
            ? throw new InvalidOperationException(Messages.DivisionByZeroMessage)
            : val1 / val2;
}